added howto use for the utility classes (#511)
ROCm/rocDecode: docs/how-to/using-rocDecode-bitstream.rst

.. meta::

>   :description: Using rocDecode with the bitstream decoder

Reviewer: Please change the `bitstream decoder` to `bitstream reader`

  :keywords: parse video, parse, decode, video decoder, video decoding, rocDecode, AMD, ROCm, bitstream decoder

********************************************************************

> Using rocDecode with the bitstream decoder

Reviewer: please change the `bitstream decoder` to `bitstream reader APIs`

********************************************************************

rocDecode core APIs are available in the |apifolder|_ of the `rocDecode GitHub repository <https://github.com/ROCm/rocDecode>`_. Utility classes that call the core APIs are available in the |utilsfolder|_ of the repository. For information about the core APIs, see :doc:`Using the rocDecode core APIs <./using-rocdecode>`.

> The `samples in the rocDecode GitHub repository <https://github.com/ROCm/rocDecode/tree/develop/samples>`_ use the utility classes. These classes are convenience classes that provide a high-level calls to the core APIs. 

Reviewer 2: ```suggestion
The `samples in the rocDecode GitHub repository <https://github.com/ROCm/rocDecode/tree/develop/samples>`_ use the utility classes. These classes are convenience classes that provide high-level calls to the core APIs. 
```

> A video stream can be decoded either with the rocDecode bitstream reader on GPU or the FFmpeg decoder on CPU. For information about using the FFmpeg decoder, see :doc:`Using rocDecode with the FFmpeg decoder <./using-rocDecode-ffmpeg>`.

Reviewer: please change this paragraph 

`A video stream can be decoded either with the rocDecode bitstream reader on GPU or the FFmpeg decoder on CPU. For information about using the FFmpeg decoder, see :doc:`Using rocDecode with the FFmpeg decoder <./using-rocDecode-ffmpeg>`.
`

to this:

`A video stream can be demultiplexed either with the rocDecode bitstream reader or the FFmpeg demultiplexer.`

> A video stream can be decoded either with the rocDecode bitstream reader on GPU or with the FFmpeg decoder on CPU. For information about using the FFmpeg decoder, see :doc:`Using rocDecode with the FFmpeg decoder <./using-rocDecode-ffmpeg>`.

Reviewer 2: ```suggestion
A video stream can be decoded either with the rocDecode bitstream reader on GPU or the FFmpeg decoder on CPU. For information about using the FFmpeg decoder, see :doc:`Using rocDecode with the FFmpeg decoder <./using-rocDecode-ffmpeg>`.
```

> A video stream can be decoded either with the rocDecode bitstream reader on GPU or with the FFmpeg decoder on CPU. For information about using the FFmpeg decoder, see :doc:`Using rocDecode with the FFmpeg decoder <./using-rocDecode-ffmpeg>`.

Reviewer 2: Is it odd that bitstream term doesn't show up until the third paragraph?
Similar with the other parallel topics. 
Perhaps it works.

Import the ``roc_bitstream_reader.h`` and ``roc_video_dec.h`` header files. These headers contain the convenience classes and functions that use the core APIs.

.. code:: C++ 

    #include "roc_bitstream_reader.h"
    #include "roc_video_dec.h"

Use the ``rocDecCreateBitstreamReader`` to get a handle to a bitstream reader. The video file's bit depth and codec ID will be returned in the ``rocDecCreateBitstreamReader`` parameters.

> From the |videodecoderaw|_ sample:

Reviewer: In the `From the |videodecoderaw|_ sample` section, please focus on explaining how the bitstream APIs are utilized in this sample. Do not include any information about the `RocVideoDecoder` utility class here. 

Additionally, we need to add a separate bullet point in the `How To Use` section, alongside the other bullet points, that conveys something similar to the following:

How to

- Use the rocDecode core APIs
- Use the rocDecode bitstream reader APIs
- Use the utility RocVideoDecoder class
- Use the utility FFMPEG demultiplexer
- Use the utility FFMPGE decoder

    *   - ``force_zero_latency``
        - ``bool``
        - Set to ``true`` to flush decoded frames in order for immediate display.

    *   - ``p_crop_rect``
        - ``const Rect *``
        - The rectangle to use for cropping. Defaults to no cropping.

    *   - ``extract_user_SEI_Message``
        - ``bool``
        - Set to ``true`` to extract Supplemental Enhancement Information (SEI) from the video stream.

    *   - ``disp_delay``
        - ``uint32_t``    
        - Delay the display by this number of frames. Defaults to 0 with no delay in displaying the frames.

    *   - ``max_width``
        - ``int``    
        - Max width. Defaults to 0.

    *   - ``max_height``
        - ``int``  
        - Max height. Defaults to 0.

    *   - ``clk_rate``
        - ``uint32_t``    
        - Clock rate. Defaults to 1000.

.. |br| raw:: html

      </br>

From ``videodecoderaw.cpp``:

.. code:: C++

    RocVideoDecoder viddec(device_id, mem_type, rocdec_codec_id, b_force_zero_latency, p_crop_rect, b_extract_sei_messages, disp_delay);

``RocVideoDecoder`` will create a parser and a decoder, and initialize HIP on the device. 

> The decoder is reused when there is a change to the video resolution without a change in the codec. When the video stream resolution changes, the existing frame buffer is deleted along with any decoded frames that are still within it, and the decoder is reconfigured for the new resolution.

Reviewer 2: ```suggestion
The decoder is reused when there's a change to the video resolution without a change in the codec. When the video stream resolution changes, the existing frame buffer is deleted along with any decoded frames that are still within it, and the decoder is reconfigured for the new resolution.
```

Reviewer 2: Should HIP be a hyperlinked?

Author: Not sure. It's not referring to the HIP, the component, but rather HIP, the backend.

To prevent the remaining frames from being deleted with the frame buffer, a callback function can be defined to save or post-process the remaining frames. 

> The |reconfig_struct|_ struct is used to store information on how to handle the reconfiguration. A callback, a user-defined flush mode, and a user-defined struct are passed to ``ReconfigParams_t``. The reconfiguration parameters are then passed to the decoder using ``SetReconfigParams``.

Reviewer 2: ```suggestion
The |reconfig_struct|_ struct stores information on how to handle the reconfiguration. A callback, a user-defined flush mode, and a user-defined struct are passed to ``ReconfigParams_t``. The reconfiguration parameters are then passed to the decoder using ``SetReconfigParams``.
```
